src/igraph_eulerian.c

```diff
@@ -0,0 +1,614 @@
+                    break;
+                }
+            }
+            
+            next = IGRAPH_TO(graph, edge) == curr ? IGRAPH_FROM(graph, edge) : IGRAPH_TO(graph, edge);
+
+            IGRAPH_CHECK(igraph_stack_push(&edge_tracker, edge));
+
+            /* remove edge here */
+            VECTOR(*outgoing_list)[curr]--;
+            VECTOR(*outgoing_list)[next]--;
+            VECTOR(visited_list)[edge] = 1;
+
+            curr = next;
+        } else { /* back track to find remaining circuit */
+            IGRAPH_CHECK(igraph_stack_push(&path, curr));
+            curr = igraph_stack_pop(&tracker);
+            if (!igraph_stack_empty(&edge_tracker)) {
+                curr_e = igraph_stack_pop(&edge_tracker);
+                IGRAPH_CHECK(igraph_stack_push(&edge_path, curr_e));
+            }
+        }
+    }
+
+    while (!igraph_stack_empty(&edge_path)) {
+        IGRAPH_CHECK(igraph_vector_push_back(res, igraph_stack_pop(&edge_path)));           
+    }
+
+    igraph_stack_destroy(&path);
+    igraph_stack_destroy(&tracker);
+    igraph_stack_destroy(&edge_path);
+    igraph_stack_destroy(&edge_tracker);
+    igraph_vector_bool_destroy(&visited_list);
+    igraph_inclist_destroy(&il);
+    IGRAPH_FINALLY_CLEAN(6);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* Hierholzer algorithm */
+int igraph_i_eulerian_path_undirected(igraph_t *graph, igraph_vector_t *path) {
+
+    /* default starting vertex is 0
+    when we have no odd vertices, we can start anywhere (usually 0)
+    when we have 2 odd vertices, we start at any one of them */
+
+    igraph_integer_t start, i;
+    igraph_inclist_t incl;
+    igraph_vector_int_t *incedges;
+    igraph_vector_t degree, out_degree;
+    igraph_bool_t has_path;
+    igraph_bool_t cycle;
+
+    has_path = 0;
+    cycle = 0;
+
+    IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+    
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
+    IGRAPH_CHECK(igraph_degree(graph, &out_degree, igraph_vss_all(), IGRAPH_OUT, IGRAPH_LOOPS));
+
+    IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+    for (i = 0; i < igraph_vector_size(&degree); i++) {
+        if (((long int) VECTOR(degree)[i]) > 0) {
+            start = i;
+            break;
+        }
+    }
+
+    igraph_vector_destroy(&degree);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &incl, IGRAPH_ALL));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &incl);
+
+    if (cycle && !has_path) {
+        IGRAPH_CHECK(igraph_i_eulerian_path_undirected_implementation(start, graph, path, &out_degree));
+    } else {
+        for (i = 0; i < igraph_vcount(graph); i++) {
+            incedges = igraph_inclist_get(&incl, i);
+            if ( igraph_vector_int_size(incedges) % 2 == 1) {
+                start = i;
+                break;
+            }
+        }
```

Comment:
Let's move the `start` calculation to one place, together with line 321 above. Then we can move the call to `igraph_i_eulerian_path_undirected_implementation` out of the `if`/`else` block, and avoid duplicating it.